zallet/src/components/json_rpc/payments.rs

```diff
@@ -0,0 +1,446 @@
+                    .iter()
+                    .any(|note| matches!(note.note().protocol(), ShieldedProtocol::Sapling))
+            }),
+            PoolType::ORCHARD => step.shielded_inputs().iter().any(|s_in| {
+                s_in.notes()
+                    .iter()
+                    .any(|note| matches!(note.note().protocol(), ShieldedProtocol::Orchard))
+            }),
+        };
+        let output_in_pool =
+            |pool_type: PoolType| step.payment_pools().values().any(|pool| *pool == pool_type);
+        let change_in_pool = |pool_type: PoolType| {
+            step.balance()
+                .proposed_change()
+                .iter()
+                .any(|c| c.output_pool() == pool_type)
+        };
+
+        let has_transparent_recipient =
+            output_in_pool(PoolType::Transparent) || change_in_pool(PoolType::Transparent);
+        let has_sapling_recipient =
+            output_in_pool(PoolType::SAPLING) || change_in_pool(PoolType::SAPLING);
+        let has_orchard_recipient =
+            output_in_pool(PoolType::ORCHARD) || change_in_pool(PoolType::ORCHARD);
+
+        if !step.transparent_inputs().is_empty() {
+            let received_addrs = step
+                .transparent_inputs()
+                .iter()
+                .map(|input| input.recipient_address())
+                .collect::<HashSet<_>>();
+
+            if received_addrs.len() > 1 {
+                if has_transparent_recipient {
+                    return PrivacyPolicy::NoPrivacy;
+                } else {
+                    return PrivacyPolicy::AllowLinkingAccountAddresses;
+                }
+            } else if has_transparent_recipient {
+                return PrivacyPolicy::AllowFullyTransparent;
+            } else {
+                return PrivacyPolicy::AllowRevealedSenders;
+            }
+        } else if has_transparent_recipient {
+            return PrivacyPolicy::AllowRevealedRecipients;
+        } else if !input_in_pool(PoolType::ORCHARD) && has_sapling_recipient
+            || !input_in_pool(PoolType::SAPLING) && has_orchard_recipient
+        {
+            // TODO: This should only trigger when there is a non-zero valueBalance.
+            return PrivacyPolicy::AllowRevealedAmounts;
+        }
+    }
+
+    // No step revealed anything.
+    PrivacyPolicy::FullPrivacy
+}
+
+pub(super) enum AddressResolutionError {
+    /// Requested [`PrivacyPolicy`] doesn’t include `AllowRevealedRecipients`.
+    TransparentRecipientNotAllowed,
+
+    /// Requested [`PrivacyPolicy`] doesn’t include `AllowRevealedRecipients`.
+    TransparentChangeNotAllowed,
+
+    /// Requested [`PrivacyPolicy`] doesn’t include `AllowRevealedAmounts`, but we don’t
+    /// have enough Sapling funds to avoid revealing amounts.
+    RevealingSaplingAmountNotAllowed,
+
+    /// Despite a lax [`PrivacyPolicy`], other factors made it impossible to find a
+    /// receiver for a recipient UA.
+    CouldNotResolveReceiver,
```

Comment:
This is similarly a `zcashd` case we don't yet handle here.

Reply:
In `zcashd` this case's error message indicated what it meant was "Orchard output with insufficient Orchard funds to avoid revealing amounts". Handled now.